File: gpboost_r/demo/boosting.R
require(gpboost)

#--------------------Simulate data----------------
# Non-linear function for simulation
f1d <- function(x) 1.7*(1/(1+exp(-(x-0.5)*20))+0.75*x)
x <- seq(from=0,to=1,length.out=200)
plot(x,f1d(x),type="l",lwd=2,col="red",main="Mean function")
# Function that simulates data. Two covariates of which only one has an effect
sim_data <- function(n){
  X=matrix(runif(2*n),ncol=2)
  # mean function plus noise
  y=f1d(X[,1])+rnorm(n,sd=0.1)
  return(list(X=X,y=y))
}
# Simulate data
n <- 1000
set.seed(1)
data <- sim_data(2 * n)
Xtrain <- data$X[1:n,]
ytrain <- data$y[1:n]
Xtest <- data$X[1:n + n,]
ytest <- data$y[1:n + n]


#--------------------Basic training using gpboost----------------
print("Train boosting model")
bst <- gpboost(data = Xtrain,
               label = ytrain,
               nrounds = 40,
               learning_rate = 0.1,
               max_depth = 6,
               min_data_in_leaf = 5,
               objective = "regression_l2",
               verbose = 0)

# You can also use an gpb.Dataset object, which stores label, data and other meta datas needed for advanced features
print("Training with gpb.Dataset")
dtrain <- gpb.Dataset(data = Xtrain, label = ytrain)
bst <- gpboost(data = dtrain,
               nrounds = 40,
               learning_rate = 0.1,
               max_depth = 6,
               min_data_in_leaf = 5,
               objective = "regression_l2",
               verbose = 0)

# Same thing using the gpb.train function
print("Training with gpb.train")
dtrain <- gpb.Dataset(data = Xtrain, label = ytrain)
bst <- gpb.train(data = dtrain,
                 nrounds = 40,
                 learning_rate = 0.1,
                 max_depth = 6,
                 min_data_in_leaf = 5,
                 objective = "regression_l2",
                 verbose = 0)

# Verbose = 1, more output
print("Train with verbose 1, print evaluation metric")
bst <- gpboost(data = dtrain,
               nrounds = 40,
               learning_rate = 0.1,
               max_depth = 6,
               min_data_in_leaf = 5,
               objective = "regression_l2",
               verbose = 1)


#--------------------Basic prediction using gpboost--------------
pred <- predict(bst, data = Xtest)
err <- mean((ytest-pred)^2)
print(paste("test-RMSE =", err))

# Compare fit to truth
x <- seq(from=0,to=1,length.out=200)
Xtest_plot <- cbind(x,rep(0,length(x)))
pred_plot <- predict(bst, data = Xtest_plot)
plot(x,f1d(x),type="l",ylim = c(-0.25,3.25), col = "red", lwd = 2,
     main = "Comparison of true and fitted value")
lines(x,pred_plot, col = "blue", lwd = 2)
legend("bottomright", legend = c("truth", "fitted"),
       lwd=2, col = c("red", "blue"), bty = "n")


#--------------------Using validation set-------------------------
# valids is a list of gpb.Dataset, each of them is tagged with a name
dtrain <- gpb.Dataset(data = Xtrain, label = ytrain)
dtest <- gpb.Dataset.create.valid(dtrain, data = Xtest, label = ytest)
valids <- list(test = dtest)

# To train with valids, use gpb.train, which contains more advanced features
# valids allows us to monitor the evaluation result on all data in the list
print("Training using gpb.train with validation data ")
bst <- gpb.train(data = dtrain,
                 nrounds = 100,
                 learning_rate = 0.1,
                 max_depth = 6,
                 min_data_in_leaf = 5,
                 objective = "regression_l2",
                 verbose = 1,
                 valids = valids,
                 early_stopping_rounds = 5)
print(paste0("Optimal number of iterations: ", bst$best_iter))

# We can change evaluation metrics, or use multiple evaluation metrics
print("Train using gpb.train with multiple validation metrics")
bst <- gpb.train(data = dtrain,
                 nrounds = 100,
                 learning_rate = 0.1,
                 max_depth = 6,
                 min_data_in_leaf = 5,
                 objective = "regression_l2",
                 verbose = 1,
                 valids = valids,
                 eval = c("l2","l1"),
                 early_stopping_rounds = 5)
print(paste0("Optimal number of iterations: ", bst$best_iter))


#--------------------Nesterov accelerated boosting-------------------------
dtrain <- gpb.Dataset(data = Xtrain, label = ytrain)
dtest <- gpb.Dataset.create.valid(dtrain, data = Xtest, label = ytest)
valids <- list(test = dtest)
print("Training using gpb.train with Nesterov acceleration")
bst <- gpb.train(data = dtrain,
                 nrounds = 100,
                 learning_rate = 0.01,
                 max_depth = 6,
                 min_data_in_leaf = 5,
                 objective = "regression_l2",
                 verbose = 1,
                 valids = valids,
                 early_stopping_rounds = 5,
                 use_nesterov_acc = TRUE)
# Compare fit to truth
x <- seq(from=0,to=1,length.out=200)
Xtest_plot <- cbind(x,rep(0,length(x)))
pred_plot <- predict(bst, data = Xtest_plot)
plot(x,f1d(x),type="l",ylim = c(-0.25,3.25), col = "red", lwd = 2,
     main = "Comparison of true and fitted value")
lines(x,pred_plot, col = "blue", lwd = 2)
legend("bottomright", legend = c("truth", "fitted"),
       lwd=2, col = c("red", "blue"), bty = "n")

File: examples/python-guide/boosting.py
# coding: utf-8
# pylint: disable = invalid-name, C0111
import gpboost as gpb
import numpy as np
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt
plt.style.use('ggplot')
print("It is recommended that the examples are run in interactive mode")

print('Simulating data...')
# Simulate and create your dataset
def f1d(x):
    """Non-linear function for simulation"""
    return (1.7 * (1 / (1 + np.exp(-(x - 0.5) * 20)) + 0.75 * x))
x = np.linspace(0, 1, 200, endpoint=True)
plt.plot(x, f1d(x), linewidth=2, color="r")
plt.title("Mean function")
plt.show()

def sim_data(n):
    """Function that simulates data. Two covariates of which only one has an effect"""
    X = np.random.rand(n, 2)
    # mean function plus noise
    y = f1d(X[:, 0]) + np.random.normal(scale=0.1, size=n)
    return ([X, y])

# Simulate data
n = 1000
data = sim_data(2 * n)
Xtrain = data[0][0:n, :]
ytrain = data[1][0:n]
Xtest = data[0][n:(2 * n), :]
ytest = data[1][n:(2 * n)]

# create dataset for gpb.train
data_train = gpb.Dataset(Xtrain, ytrain)
data_eval = gpb.Dataset(Xtest, ytest, reference=data_train)

# specify your configurations as a dict
params = {
    'objective': 'regression_l2',
    'metric': {'l2', 'l1'},
    'learning_rate': 0.1,
    'max_depth': 6,
    'min_data_in_leaf': 5,
    'verbose': 0
}

print('Starting training...')
# train
evals_result = {}  # record eval results for plotting
bst = gpb.train(params=params,
                train_set=data_train,
                num_boost_round=100,
                valid_sets=data_eval,
                early_stopping_rounds=5,
                evals_result = evals_result)

# plot validation scores
gpb.plot_metric(evals_result, metric='l1', figsize=(10, 5))
plt.show()

print('Saving model...')
# save model to file
bst.save_model('model.txt')

print('Starting predicting...')
# predict
y_pred = bst.predict(Xtest, num_iteration=bst.best_iteration)
# eval
print('The rmse of prediction is:', mean_squared_error(ytest, y_pred) ** 0.5)

# compare fit to truth
Xtest_plot = np.zeros((200, 2))
Xtest_plot[:, 0] = np.linspace(0, 1, 200, endpoint=True)
y_pred_plot = bst.predict(Xtest_plot, num_iteration=bst.best_iteration)
plt.plot(Xtest_plot[:, 0], y_pred_plot, linewidth=2, color="b", label="fit")
plt.plot(x, f1d(x), linewidth=2, color="r", label="true")
plt.title("Comparison of true and fitted value")
plt.legend()
plt.show()

print('Starting training with custom evaluation function...')
# self-defined eval metric
# f(preds: array, train_data: Dataset) -> name: string, eval_result: float, is_higher_better: bool
# quantile loss
def quantile_loss(preds, train_data):
    alpha = 0.95
    labels = train_data.get_label()
    diff = labels - preds
    dummy = diff < 0
    return 'quantile_loss', np.mean((alpha - dummy) * diff), False

params = {
    'objective': 'regression_l2',
    'metric': 'quantile_loss',
    'learning_rate': 0.1,
    'max_depth': 6,
    'min_data_in_leaf': 5,
    'verbose': 0
}

bst = gpb.train(params=params,
                train_set=data_train,
                num_boost_round=100,
                feval=quantile_loss,
                valid_sets=data_eval,
                early_stopping_rounds=5)


# Several custom evaluation functions
# l4 loss
def l4_loss(preds, train_data):
    labels = train_data.get_label()
    return 'l4_loss', np.mean((preds - labels) ** 4), False


bst = gpb.train(params=params,
                train_set=data_train,
                num_boost_round=100,
                feval=lambda preds, train_data: [quantile_loss(preds, train_data),
                                                 l4_loss(preds, train_data)],
                valid_sets=data_eval,
                early_stopping_rounds=5)

# feature importances
bst = gpb.train(params=params,
                train_set=data_train,
                num_boost_round=20)

print('Feature importances:', list(bst.feature_importance()))

print('Plotting feature importances...')
ax = gpb.plot_importance(bst, max_num_features=10)
plt.show()

print('Plotting split value histogram...')
ax = gpb.plot_split_value_histogram(bst, feature='Column_0', bins='auto')
plt.show()

print('Plotting 10th tree...')  # one tree use categorical feature to split
ax = gpb.plot_tree(bst, tree_index=10, figsize=(15, 15), show_info=['split_gain'])
plt.show()

print('Starting training Boosting with Nesterov accelaration...')
# Boosting with Nesterov acceleration
params = {
    'objective': 'regression_l2',
    'metric': {'l2'},
    'learning_rate': 0.1,
    'max_depth': 6,
    'min_data_in_leaf': 5,
    'verbose': 1,
    'use_nesterov_acc': True
}
bst = gpb.train(params=params,
                train_set=data_train,
                num_boost_round=100,
                valid_sets=data_eval,
                early_stopping_rounds=5)
# predict
y_pred = bst.predict(Xtest, num_iteration=bst.best_iteration)
# eval
print('The rmse of prediction is:', mean_squared_error(ytest, y_pred) ** 0.5)
